File: android-ui/home/src/androidMain/kotlin/io/github/viacheslav/chugunov/moniqo/ui/home/screen/HomeViewModel.kt
package io.github.viacheslav.chugunov.moniqo.ui.home.screen

import androidx.lifecycle.viewModelScope
import io.github.viacheslav.chugunov.moniqo.core.model.DealRanges
import io.github.viacheslav.chugunov.moniqo.core.model.RatePair
import io.github.viacheslav.chugunov.moniqo.core.usecase.GetDealRangesFlowUseCase
import io.github.viacheslav.chugunov.moniqo.core.usecase.GetRatePairFlowUseCase
import io.github.viacheslav.chugunov.moniqo.core.usecase.SaveFromRateUseCase
import io.github.viacheslav.chugunov.moniqo.core.usecase.SaveToRateUseCase
import io.github.viacheslav.chugunov.moniqo.ui.core.AppViewModel
import kotlinx.coroutines.flow.combineTransform
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import kotlin.text.iterator

internal class HomeViewModel(
    private val getRatePairFlowUseCase: GetRatePairFlowUseCase,
    private val saveFromRateUseCase: SaveFromRateUseCase,
    private val saveToRateUseCase: SaveToRateUseCase,
    private val getDealRangesFlowUseCase: GetDealRangesFlowUseCase,
    private val mapper: HomeMapper,
) : AppViewModel<HomeState, HomeIntent, HomeEffect>(HomeState.Loading) {
    private var currentRatePair: RatePair? = null
    private var currentDealRanges: DealRanges? = null

    init {
        combineTransform<DealRanges, RatePair, HomeState>(
            getDealRangesFlowUseCase(),
            getRatePairFlowUseCase()
        ) { dealRanges, ratePair ->
            currentRatePair = ratePair
            currentDealRanges = dealRanges
            emit(
                mapper.toHomeState(
                    ratePair = ratePair,
                    dealRanges = dealRanges,
                    currentContentState = childState()
                )
            )
        }.onEach { state ->
            updateState { state }
        }.launchIn(viewModelScope)
    }

    override fun onIntent(intent: HomeIntent) {
        when (intent) {
            is HomeIntent.ChangeFromAmount -> handleChangeFromAmount(intent.input)
            is HomeIntent.ChangeToAmount -> handleChangeToAmount(intent.input)
            is HomeIntent.SwapCurrencies -> handleSwapCurrencies()
        }
    }

    private fun handleChangeFromAmount(input: String) {
        val sanitized = sanitize(input)
        val pair = currentRatePair
        val dealRanges = currentDealRanges ?: return
        updateContent { current ->
            if (pair == null) {
                current.copy(fromAmount = sanitized)
            } else {
                val content = childState<HomeState.Content>()?.copy(fromAmount = input)
                mapper.toHomeState(pair, dealRanges, content)
            }
        }
    }

    private fun handleChangeToAmount(input: String) {
        val sanitized = sanitize(input)
        val pair = currentRatePair
        val dealRanges = currentDealRanges ?: return
        updateContent { current ->
            current.copy(
                toAmount = sanitized,
                analysis =
                    mapper.toExchangeAnalysis(
                        current.fromAmount,
                        sanitized,
                        pair?.let { mapper.toOfficialRate(it) },
                        current.fromCurrency,
                        current.toCurrency,
                        dealRanges,
                    ),
            )
        }
    }

    private fun handleSwapCurrencies() {
        val pair = currentRatePair ?: return
        viewModelScope.launch {
            saveFromRateUseCase(pair.toRate)
            saveToRateUseCase(pair.fromRate)
        }
    }

    private fun updateContent(block: (HomeState.Content) -> HomeState.Content) {
        updateState { if (it is HomeState.Content) block(it) else it }
    }

    private fun sanitize(input: String): String =
        buildString {
            var hasDot = false
            for (char in input) {
                when {
                    char.isDigit() -> append(char)
                    char == '.' && !hasDot -> {
                        hasDot = true
                        append(char)
                    }
                }
            }
        }
}

File: iosApp/iosApp/home/screen/HomeViewModel.swift
import Foundation
import Shared

final class HomeViewModel: ObservableObject {
    @Published private(set) var state: HomeState = .loading

    private let mapper: HomeMapper
    private var ratePair: RatePair?
    private var currentGoodDealMax: Int = 5
    private var currentMediumDealMax: Int = 10
    private var stopObservation: (() -> Void)?
    private var stopRangesObservation: (() -> Void)?
    private var localeObservation: NSObjectProtocol?
    private var pendingJobs: [() -> Void] = []

    init(mapper: HomeMapper) {
        self.mapper = mapper
        startObservingRatePair()
        startObservingDealRanges()
        localeObservation = NotificationCenter.default.addObserver(
            forName: .appLocaleDidChange,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let self, let pair = self.ratePair else { return }
            self.state = .content(self.mapper.toHomeContent(
                pair: pair,
                fromAmount: self.currentFromAmount,
                goodDealMax: self.currentGoodDealMax,
                mediumDealMax: self.currentMediumDealMax
            ))
        }
    }

    deinit {
        stopObservation?()
        stopRangesObservation?()
        pendingJobs.forEach { $0() }
        localeObservation.map { NotificationCenter.default.removeObserver($0) }
    }

    func onIntent(_ intent: HomeIntent) {
        switch intent {
        case .changeFromAmount(let input):
            handleChangeFromAmount(input)
        case .changeToAmount(let input):
            handleChangeToAmount(input)
        case .swapCurrencies:
            guard let pair = ratePair else { return }
            pendingJobs.append(HomeUseCaseBridgeKt.swapRates(fromRate: pair.toRate, toRate: pair.fromRate))
        }
    }

    private func startObservingRatePair() {
        stopObservation = HomeUseCaseBridgeKt.observeRatePair { [weak self] pair in
            guard let self else { return }
            self.ratePair = pair
            self.state = .content(self.mapper.toHomeContent(
                pair: pair,
                fromAmount: self.currentFromAmount,
                goodDealMax: self.currentGoodDealMax,
                mediumDealMax: self.currentMediumDealMax
            ))
        }
    }

    private func startObservingDealRanges() {
        stopRangesObservation = SettingsBridgeKt.observeDealRanges { [weak self] ranges in
            guard let self else { return }
            self.currentGoodDealMax = Int(ranges.good)
            self.currentMediumDealMax = Int(ranges.medium)
            guard let pair = self.ratePair else { return }
            self.state = .content(self.mapper.toHomeContent(
                pair: pair,
                fromAmount: self.currentFromAmount,
                goodDealMax: self.currentGoodDealMax,
                mediumDealMax: self.currentMediumDealMax
            ))
        }
    }

    private var currentFromAmount: String {
        if case .content(let content) = state { return content.fromAmount }
        return ""
    }

    private func handleChangeFromAmount(_ input: String) {
        let sanitized = sanitize(input)
        guard let pair = ratePair else {
            if case .content(var content) = state {
                content.fromAmount = sanitized
                state = .content(content)
            }
            return
        }
        state = .content(mapper.toHomeContent(
            pair: pair,
            fromAmount: sanitized,
            goodDealMax: currentGoodDealMax,
            mediumDealMax: currentMediumDealMax
        ))
    }

    private func handleChangeToAmount(_ input: String) {
        let sanitized = sanitize(input)
        guard case .content(var content) = state else { return }
        let officialRate = ratePair.map { mapper.toOfficialRate(pair: $0) }
        content.toAmount = sanitized
        content.analysis = mapper.toExchangeAnalysis(
            fromAmount: content.fromAmount,
            toAmount: sanitized,
            officialRate: officialRate,
            fromCurrency: content.fromCurrency,
            toCurrency: content.toCurrency,
            goodDealMax: currentGoodDealMax,
            mediumDealMax: currentMediumDealMax
        )
        state = .content(content)
    }

    private func sanitize(_ input: String) -> String {
        var hasDot = false
        var result = ""
        for char in input {
            if char.isNumber {
                result.append(char)
            } else if char == "." && !hasDot {
                hasDot = true
                result.append(char)
            }
        }
        return result
    }
}
